layers/shader_attribute_validation.h

```diff
@@ -0,0 +1,308 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Author: Arno Galvez <arno@lunarg.com>
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <array>
+
+#include "shader_module.h"
+#include "range_vector.h"
+
+class ShaderAttributeValidation {
+  public:
+    enum class ComponentStatus {
+        Uninitialized,
+        Seen,
+        Unseen,
+    };
+
+    ShaderAttributeValidation(const SHADER_MODULE_STATE &shader_module,
+                              const std::map<location_t, interface_var>::const_iterator &shader_attributes_iterator,
+                              const std::map<location_t, interface_var>::const_iterator &shader_attributes_iterator_end)
+        : is_valid_(shader_attributes_iterator != shader_attributes_iterator_end) {
+        if (is_valid_) {
+            start_location_ = shader_attributes_iterator->first.first;
+            start_component_ = shader_attributes_iterator->first.second;
+            current_global_component_ = start_location_ * 4 + start_component_;
+
+            const auto def = shader_module.FindDef(shader_attributes_iterator->second.type_id);
+            const auto base_type_components_count = shader_module.GetNumComponentsInBaseType(def);
+            const auto total_byte_size = shader_module.GetTypeBytesSize(def);
+            const auto array_size = shader_module.GetArraySize(def);  // Will be 1 if not an array
+            assert(array_size > 0);
+            // a 32 bit or less component count as 1, a 64 bit component as 2.
+            if (total_byte_size == 2 * base_type_components_count * 4 * array_size) {
+                components_count_ = base_type_components_count * 2;
+            } else {
+                components_count_ = base_type_components_count;
+            }
+            assert(components_count_ > 0);
+            assert(components_count_ <= static_cast<uint32_t>(components_.size()));
+            for (uint32_t component_i = 0; component_i < components_count_; ++component_i) {
+                components_[component_i] = ComponentStatus::Unseen;
+            }
+            current_components_left_ = components_count_;
+
+            interface_ = &shader_attributes_iterator->second;
+        }
+    }
+
+    static void TagMatchingComponentsAsSeen(ShaderAttributeValidation &lhs, ShaderAttributeValidation &rhs) {
+        const sparse_container::range<uint32_t> lhs_range(lhs.current_global_component_,
+                                                          lhs.current_global_component_ + lhs.current_components_left_);
+        const sparse_container::range<uint32_t> rhs_range(rhs.current_global_component_,
+                                                          rhs.current_global_component_ + rhs.current_components_left_);
+        const sparse_container::range<uint32_t> range_intersection = lhs_range & rhs_range;
+        if (range_intersection.non_empty()) {
```

Comment:
Return early to avoid indenting the majority of the logic.

```suggestion
        if (range_intersection.empty()) { return; }
```